### PR TITLE
RavenDB-18820 Corax - growable buffer for analyzers

### PR DESCRIPTION
**Pull request by MattAszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18820
https://issues.hibernatingrhinos.com/issue/RavenDB-18824

### Additional description

RavenDB-18820

Analyzer requires a buffer before doing work. We don't want to allocate new buffer for each field or to allocate too much so we allocate default buffer. When field is bigger than buffer we just extend buffer.

RavenDB-18824

Changing the approach in dealing with complex objects in Corax. Instead of immediately throwing we'll default to `Indexing: No` for static indexes (unless it was defined otherwise in an index definition)

### Type of change

- Bug fix
- Optimization

### How risky is the change?
- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/Constants.cs [hunk 1]
```diff
@@ -80,6 +80,11 @@ public enum Operator
             }
         }
 
+        public static class Analyzers
+        {
+            public const int DefaultBufferForAnalyzers = 4 * Size.Kilobyte; //is this enough?
```
Comment (ayende) [inline src/Corax/Constants.cs]: That should be plenty big enough for most scenarios, yes.

File: src/Corax/IndexSearcher/IndexSearcher.cs [hunk 2]
```diff
@@ -139,15 +138,17 @@ internal unsafe ReadOnlySpan<byte> ApplyAnalyzer(ReadOnlySpan<byte> originalTerm
...
-        
-        Span<byte> encoded = new byte[outputSize];
-        Token* tokensPtr = stackalloc Token[tokenSize];
-        var tokens = new Span<Token>(tokensPtr, tokenSize);
-        
-        analyzer.Execute(originalTerm, ref encoded, ref tokens);
-        Debug.Assert(tokens.Length == 1, $"{nameof(ApplyAnalyzer)} should create only 1 token as a result.");
 
-        return encoded;
+        var buffer = ArrayPool<byte>.Shared.Rent(outputSize); // We can release it only on dispose. todo maciej: write stack of arrays to return.
```
Comment (ayende) [inline src/Corax/IndexSearcher/IndexSearcher.cs]: this is now leaked, no?

File: src/Corax/IndexWriter.cs [hunk 3]
```diff
@@ -504,6 +508,19 @@ public bool TryDeleteEntry(string key, string term)
             return true;
         }
 
+        private void UnlikelyGrowBuffer(out Span<byte> buffer, out Span<Token> tokens, int newBufferSize, int newTokenSize)
+        {
+            Debug.Assert(buffer.Length <= newBufferSize);
```
Comment (ayende) [inline src/Corax/IndexWriter.cs:517]: That is valid? It is an out param - it _has_ no length.

Comment (maciejaszyk) [reply src/Corax/IndexWriter.cs:517]: Good catch

File: src/Corax/IndexSearcher/IndexSearcher.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Corax/IndexSearcher/IndexSearcher.cs]: Yes, fixed now.

**Comment by arekpalinski** [comment]
Please check failing Debug.Assert

**Comment by maciejaszyk** [comment]
> Please check failing Debug.Assert

fixed

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**